src/atom_vec.cpp

```diff
@@ -2330,19 +2330,19 @@ void AtomVec::write_data_restore_restricted()
   // no other write_data Velocities fields are Nx3 double arrays
 
   if (v_hold) {
-    memcpy(&v[0][0],&v_hold[0][0],3*nlocal*sizeof(double));
+    memcpy(&v[0][0],&v_hold[0][0],(3*nlocal*sizeof(double))&MEMCPYMASK);
```

Comment:
just a Q about this - say nlocal is only a bit smaller than MAXINT (obviously unlikely) - does 3*nlocal
overflow before the multiply by sizeof(double) ?

Reply:
> does 3*nlocal overflow before the multiply by sizeof(double) ?

Looks like it. Check out: https://godbolt.org/z/WPzzrThPz

Reply:
So you're saying the MEMCPYMASK will prevent all the (bogus) compiler warnings?
But we should also (not in this PR) make sure we are not potentially causing these kinds of overflows?
Either by ordering the multiplies small to large, or by assigning the arg to a bigint (with appropriate casting)
and using the bigint as the arg?
We do the latter in other places in the code already, but undoubtably not everywhere.

Reply:
> So you're saying the MEMCPYMASK will prevent all the (bogus) compiler warnings?

It does for me.  ...and it is independent of the overflow issue. It just reduces the possible size from 64-bits to 63-bits which GCC for unknown reasons has decided is the allowed maximum.

> make sure we are not potentially causing these kinds of overflows?

I am genuinely surprised it hasn't shown with any of the static code analysis tools we use (Coverity Scan, CodeQL, or compiler warnings with `-Wall -Wextra`). There were numerous overflow warnings that I have fixed over the years. Most often inconsequential ones, like in the `memory_usage()` functions. There are over 200 files where there is now a cast to double for the first number:
```
lammps/src$ grep -l bytes.*+=.*double *.cpp */*.cpp | wc -l
203
```

> Either by ordering the multiplies small to large, or by assigning the arg to a bigint (with appropriate casting)
and using the bigint as the arg?

Without parenthesis, expressions are evaluated left to right, so the simplest is to cast the first number or variable to the final data type. 

> We do the latter in other places in the code already, but undoubtably not everywhere.

There is too much code to check, so it needs some tool to automate this. Something like clang-tidy or a compiler warning should do. But then again, there are use cases, where the wrap around, specifically for integers is desired, so the check has to be more narrow, e.g. when passing something to a function that accepts a size_t argument or similar.